docs/reference/troubleshooting/corruption-issues.asciidoc

```diff
@@ -0,0 +1,83 @@
+end and then never modified or overwritten. This access pattern means the
+checksum computation is very simple and can happen on-the-fly as the file is
+written, and also makes it very unlikely that an incorrect checksum is due to a
+userspace bug at the time the file was written. The routine that computes the
+checksum is straightforward, widely used, and very well-tested, so you can be
+very confident that a checksum mismatch really does indicate that the data read
+from disk is different from the data that {es} previously wrote.
+
+The files that make up a Lucene index are written in full before they are used.
+If a file is needed to recover an index after a restart then your storage
+system will previously have confirmed to {es} that this file was durably synced
+to disk. On Linux this means that the `fsync()` system call returned
+successfully. {es} sometimes detects that an index is corrupt because a file
+needed for recovery has been truncated or is missing its footer. This indicates
+that your storage system acknowledges durable writes incorrectly.
+
+There are many possible explanations for {es} detecting corruption in your
+cluster. Databases like {es} generate a challenging I/O workload that may find
+subtle infrastructural problems which other tests may miss. {es} is known to
+expose the following problems as file corruptions:
+
+- filesystem bugs, especially in newer and nonstandard filesystems which might
+  not have seen enough real-world production usage to be confident that they
+work correctly.
+
+- [kernel
+  bugs]https://www.elastic.co/blog/canonical-elastic-and-google-team-up-to-prevent-data-corruption-in-linux
+
+- bugs in firmware running on the drive or RAID controller
+
+- incorrect configuration, for instance configuring `fsync()` to report success
+  before all durable writes have completed
+
+- faulty hardware (for instance the drive, RAID controller, RAM or CPU)
+
+Data corruption typically doesn't result in other evidence of problems apart
+from the checksum mismatch. Do not interpret this as an indication that your
+storage subsystem is working correctly and therefore that {es} itself caused
+the corruption. It is rare for faulty storage to show any evidence of problems
+apart from the data corruption, but data corruption itself is a very strong
+indicator that your storage subsystem is not working correctly.
+
+To narrow down the source of the corruptions, systematically change components
+in your cluster until the corruptions stop. For instance, to rule out {es} as
+the source of data corruption, generate an I/O workload using something other
+than {es} and look for data integrity errors. On Linux the `fio` and
+`stress-ng` tools can both generate challenging I/O workloads and verify the
+integrity of the data they write. You can check that durable writes persist
+across power outages using a script such as
+[`diskchecker.pl`]https://gist.github.com/bradfitz/3172656. Try different
```

Comment:
Unclear if we should mention these tools like this. Maybe we should give example invocations for `fio` and `stress-ng`? `diskchecker.pl` looks janky but it's pretty much the best way to find `fsync()` bugs (it's in the [PostgreSQL docs](https://www.postgresql.org/docs/current/wal-reliability.html)).